src/Raven.Server/Documents/Indexes/Index.cs

```diff
@@ -380,6 +382,9 @@ protected void ExecuteIndexing()
 
                             DoIndexingWork(stats, cts.Token);
 
+                            _indexingBatchCompleted.SetByAsyncCompletion();
+                            _indexingBatchCompleted.Reset();
+
```

Comment:
We can introduce a method which would do basically the same, however in a different order:
1) replace TCS by a new instance
2) mark the previous instance of TCS as completed

Then such method would be atomic.